### PR TITLE
Add `exec_opt`, which is similar to exec, but returns an option.

### PR DESCRIPTION
**Pull request by Drup**
I fail to see how neither @c-cube nor @rgrinberg added that already. :D

### CODE REVIEW

File: lib/re.ml [hunk 1]
```diff
@@ -947,6 +947,13 @@ let exec ?(pos = 0) ?(len = -1) re s =
     Match substr -> substr
   | _            -> raise Not_found
 
+let exec_opt ?(pos = 0) ?(len = -1) re s =
+  if pos < 0 || len < -1 || pos + len > String.length s then
+    invalid_arg "Re.exec";
```
Comment (dsheets) [inline lib/re.ml]: `Re.exec_opt`

**Comment by dsheets** [comment]
Also, why don't these functions share their body and one use the other?

**Comment by Drup** [comment]
I have factorized the body of all the exec functions. I also added `get_start/end` in a second commit, which I found to make the code quite clearer.

**Comment by rgrinberg** [comment]
I personally haven't added it because I would prefer to have `exec_opt` as the default `exec` and have the old `exec` be `exec_exn`. But I guess breaking compatibility is probably not going to be possible.

As for your last commit, doesn't it conflict with #43? It's my preference to move all match accessors to a `Match` sub module. It would always shorten up `get_{start,end}` to something like `{start, end_}` which is IMO a little more pleasing.

**Comment by Drup** [comment]
I don't think changing the type of `Re.exec` is possible at this point without breaking a lot of code.

I agree, a `Match` module is a better idea.

**Comment by rgrinberg** [comment]
@Drup do you mind putting the first commit in a separate PR? I think it's ok to merge.

**Comment by Drup** [comment]
@rgrinberg I droped the second commit.

**Pull request merged**